File: app/src/main/java/com/sasfmlzr/findfilm/fragment/discoverfilm/DiscoverFilmViewModel.java
package com.sasfmlzr.findfilm.fragment.discoverfilm;

import android.databinding.BaseObservable;
import android.databinding.Bindable;
import android.support.v7.widget.RecyclerView;

import com.android.databinding.library.baseAdapters.BR;
import com.sasfmlzr.findfilm.adapter.DiscoverRecyclerAdapter;
import com.sasfmlzr.findfilm.model.RetrofitSingleton;
import com.sasfmlzr.findfilm.request.DiscoverMovieRequest;
import com.sasfmlzr.findfilm.request.FindFilmApi;

import java.util.List;
import java.util.Objects;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static com.sasfmlzr.findfilm.model.SystemSettings.API_KEY;
import static com.sasfmlzr.findfilm.model.SystemSettings.LANGUAGE;

public class DiscoverFilmViewModel extends BaseObservable {
    @Bindable
    private RecyclerView.Adapter recyclerViewAdapter;
    private DiscoverFilmFragment.OnFilmSelectedListener filmSelectedListener;
    private String querySearch;
    private int countLoadedPages = 1;
    private boolean isFirstList = true;

    DiscoverFilmViewModel(DiscoverFilmFragment.OnFilmSelectedListener filmSelectedListener) {
        this.filmSelectedListener = filmSelectedListener;
    }

    public void runRequest(String querySearch) {
        isFirstList = true;
        this.querySearch = querySearch;

        if (querySearch == null) {
            runRequestFilm(1, filmListListener());
        } else {
            DiscoverFilmFragment.SearchCallback callback = this::setAdapterDiscoverFilm;
            runSearchRequestFilm(querySearch, callback);
        }
    }

    public void runSearchRequestFilm(String query, DiscoverFilmFragment.SearchCallback callback) {
        FindFilmApi findFilmApi = RetrofitSingleton.getFindFilmApi();
        findFilmApi.getSearchMovie(API_KEY, LANGUAGE, query, 1)
                .enqueue(new Callback<DiscoverMovieRequest>() {
                    @Override
                    public void onResponse(Call<DiscoverMovieRequest> call, Response<DiscoverMovieRequest> response) {
                        DiscoverMovieRequest discoverMovieRequest = response.body();
                        callback.isFind(discoverMovieRequest.getResults());
                    }

                    @Override
                    public void onFailure(Call<DiscoverMovieRequest> call, Throwable t) {
                    }
                });
    }

    public RecyclerView.Adapter getRecyclerViewAdapter() {
        return recyclerViewAdapter;
    }

    private void setRecyclerViewAdapter(RecyclerView.Adapter recyclerViewAdapter) {
        this.recyclerViewAdapter = recyclerViewAdapter;
        notifyPropertyChanged(BR.recyclerViewAdapter);
    }

    private void runRequestFilm(int countLoadedPages, DiscoverFilmFragment.FilmListComplete callback) {
        FindFilmApi findFilmApi = RetrofitSingleton.getFindFilmApi();
        findFilmApi.getDiscoverMovie(API_KEY, LANGUAGE, countLoadedPages)
                .enqueue(new Callback<DiscoverMovieRequest>() {
                    @Override
                    public void onResponse(Call<DiscoverMovieRequest> call,
                                           Response<DiscoverMovieRequest> response) {
                        callback.isCompleted(response.body().getResults());
                    }

                    @Override
                    public void onFailure(Call<DiscoverMovieRequest> call, Throwable t) {

                    }
                });
    }

    private DiscoverFilmFragment.FilmListComplete filmListListener() {
        return (filmList) -> {
            setAdapterDiscoverFilm(filmList);
            countLoadedPages++;
        };
    }

    private void setAdapterDiscoverFilm(List<DiscoverMovieRequest.Result> filmList) {
        if (isFirstList) {
            DiscoverFilmFragment.RecyclerElementEnded callback = () -> {
                if (querySearch == null) {
                    runRequestFilm(countLoadedPages, filmListListener());
                }
            };
            RecyclerView.Adapter adapter =
                    new DiscoverRecyclerAdapter(filmList, filmSelectedListener, callback);

            setRecyclerViewAdapter(adapter);
            isFirstList = false;
        } else {
            ((DiscoverRecyclerAdapter) Objects.requireNonNull(
                    getRecyclerViewAdapter()))
                    .addElements(filmList);
        }
    }
}

File: app/src/main/java/com/sasfmlzr/findfilm/fragment/discoverfilm/DiscoverFilmViewModel.kt
package com.sasfmlzr.findfilm.fragment.discoverfilm

import androidx.databinding.BaseObservable
import androidx.databinding.Bindable
import androidx.databinding.library.baseAdapters.BR
import androidx.recyclerview.widget.RecyclerView
import com.sasfmlzr.findfilm.adapter.DiscoverRecyclerAdapter
import com.sasfmlzr.findfilm.model.RetrofitSingleton
import com.sasfmlzr.findfilm.model.SystemSettings
import com.sasfmlzr.findfilm.request.DiscoverMovieRequest
import com.sasfmlzr.findfilm.request.FindFilmApi
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*

class DiscoverFilmViewModel internal constructor(
    private val filmSelectedListener: DiscoverFilmFragment.OnFilmSelectedListener?
) : BaseObservable() {

    @Bindable
    var recyclerViewAdapter: RecyclerView.Adapter<*>? = null
        private set(recyclerViewAdapter) {
            field = recyclerViewAdapter
            notifyPropertyChanged(BR.recyclerViewAdapter)
        }
    private var querySearch: String? = null
    private var countLoadedPages = 1
    private var isFirstList = true

    fun runRequest(querySearch: String?) {
        isFirstList = true
        this.querySearch = querySearch
        if (querySearch == null) {
            runRequestFilm(1, filmListListener())
        } else {
            val callback =
                object : DiscoverFilmFragment.SearchCallback {
                    override fun isFind(filmList: List<DiscoverMovieRequest.Result>) {
                        setAdapterDiscoverFilm(filmList)
                    }
                }
            runSearchRequestFilm(querySearch, callback)
        }
    }

    fun runSearchRequestFilm(query: String?, callback: DiscoverFilmFragment.SearchCallback) {
        val findFilmApi: FindFilmApi = RetrofitSingleton.findFilmApi!!
        findFilmApi.getSearchMovie(SystemSettings.API_KEY, SystemSettings.LANGUAGE, query, 1)
            .enqueue(object : Callback<DiscoverMovieRequest?> {
                override fun onResponse(
                    call: Call<DiscoverMovieRequest?>,
                    response: Response<DiscoverMovieRequest?>
                ) {
                    val discoverMovieRequest: DiscoverMovieRequest? = response.body()
                    callback.isFind(discoverMovieRequest?.results!!)
                }

                override fun onFailure(call: Call<DiscoverMovieRequest?>, t: Throwable) {}
            })
    }

    private fun runRequestFilm(
        countLoadedPages: Int,
        callback: DiscoverFilmFragment.FilmListComplete
    ) {
        val findFilmApi: FindFilmApi = RetrofitSingleton.findFilmApi!!
        findFilmApi.getDiscoverMovie(
            SystemSettings.API_KEY,
            SystemSettings.LANGUAGE,
            countLoadedPages
        )
            .enqueue(object : Callback<DiscoverMovieRequest> {
                override fun onResponse(
                    call: Call<DiscoverMovieRequest>,
                    response: Response<DiscoverMovieRequest>
                ) {
                    callback.isCompleted(response.body()?.results!!)
                }

                override fun onFailure(call: Call<DiscoverMovieRequest>, t: Throwable) {}
            })
    }

    private fun filmListListener(): DiscoverFilmFragment.FilmListComplete {
        return object : DiscoverFilmFragment.FilmListComplete {
            override fun isCompleted(filmList: List<DiscoverMovieRequest.Result>) {
                setAdapterDiscoverFilm(filmList)
                countLoadedPages++
            }
        }
    }

    private fun setAdapterDiscoverFilm(filmList: List<DiscoverMovieRequest.Result>) {
        if (isFirstList) {
            val callback = object : DiscoverFilmFragment.RecyclerElementEnded {
                override val isEnded: Unit
                    get() = if (querySearch == null) {
                        runRequestFilm(countLoadedPages, filmListListener())
                    } else Unit

            }
            val adapter: RecyclerView.Adapter<*> =
                DiscoverRecyclerAdapter(filmList.toMutableList(), filmSelectedListener!!, callback)
            recyclerViewAdapter = adapter
            isFirstList = false
        } else {
            (Objects.requireNonNull(
                recyclerViewAdapter
            ) as DiscoverRecyclerAdapter)
                .addElements(filmList)
        }
    }
}
